File: 1379-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/1379-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        return solve(cloned, target -> val);
    }
private:
    TreeNode* solve(TreeNode* root, int target) {
        if(root == nullptr)
            return nullptr;
        if(root -> val == target)
            return root;
        
        auto left = solve(root -> left, target);
        auto right = solve(root -> right, target);
        
        if(left == nullptr)
            return right;
        return left;
    }
};

File: 1379-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/1379-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def solve(self, root, target):
        if root == None:
            return root
        
        if root.val == target:
            return root
        
        left = self.solve(root.left, target)
        right = self.solve(root.right, target)
        
        if left == None:
            return right
        return left
    
    def daily(self, rootOriginal, rootCloned, target):
        if rootOriginal == None:
            return rootOriginal
        
        if rootOriginal == target:
            return rootCloned
        
        left = self.daily(rootOriginal.left, rootCloned.left, target)
        right = self.daily(rootOriginal.right, rootCloned.right, target)
        
        if left == None:
            return right
        return left
        
        
    def getTargetCopy(self, original: TreeNode, cloned: TreeNode, target: TreeNode) -> TreeNode:
        return self.daily(original, cloned, target)
        # return self.solve(cloned, target.val)